cmd/snap/cmd_run.go

```diff
@@ -530,6 +531,62 @@ func migrateXauthority(info *snap.Info) (string, error) {
 	return targetPath, nil
 }
 
+func activateXdgDocumentPortal(info *snap.Info) error {
+	// Don't do anything for snaps that don't plug the desktop interface
+	//
+	// NOTE: This check is imperfect because we don't really know
+	// if the interface is connected or not but this is an
+	// acceptable compromise for not having to communicate with
+	// snapd in snap run. In a typical desktop session the
+	// document portal can be in use by many applications, not
+	// just by snaps, so this is at most, pre-emptively using some
+	// extra memory.
+	plugsDesktop := false
+	for _, plug := range info.Plugs {
+		if plug.Interface == "desktop" {
+			plugsDesktop = true
+			break
+		}
+	}
+	if !plugsDesktop {
+		return nil
+	}
+
+	// If there is no session bus, our job is done.  We check this
+	// manually to avoid dbus.SessionBus() auto-launching a new
+	// bus.
+	busAddress := osGetenv("DBUS_SESSION_BUS_ADDRESS")
+	if len(busAddress) == 0 {
+		return nil
+	}
+
+	conn, err := dbus.SessionBus()
+	if err != nil {
+		return err
+	}
+
+	portal := conn.Object("org.freedesktop.portal.Documents",
+		"/org/freedesktop/portal/documents")
```

Comment:
If the name is not currently owned, the bus will check whether it has any configuration to activate the name.  If there is no such configuration (because portals are not installed on the system), then the bus will immediately return a `ServiceUnknown` error.

With portals installed dbus-daemon will launch `xdg-document-portal`, wait for the name to be claimed on the bus and forward the method call.

The requirement here is for the document portal mount point to exist before we set up the mount namespace for the confined application.  Even once we've got persistent user mount namespaces, I'd be hesitant to remove this code, since the time between xdg-document-portal starting up and whatever service reacts by patching the mount namespace would represent a window when the confined app could see every file published through the portal.

Managing to retain a single directory file descriptor to the base of the document portal file system would be enough, together with `openat`, `fdopendir`, etc.

Reply:
Okay, if we nail the previous point in a really nice way, perhaps we can make it a NOOP for all the cases where it will be irrelevant.